swarm-test/src/lib.rs

```diff
@@ -0,0 +1,183 @@
+                            dialer_done = true;
+                        }
+                        other => {
+                            log::debug!("Ignoring {:?}", other);
+                        }
+                    }
+                },
+                listener_event = other.select_next_some() => {
+                    match listener_event {
+                        SwarmEvent::ConnectionEstablished { .. } => {
+                            listener_done = true;
+                        }
+                        SwarmEvent::IncomingConnectionError { error, .. } => {
+                            panic!("Failure in incoming connection {}", error);
+                        }
+                        other => {
+                            log::debug!("Ignoring {:?}", other);
+                        }
+                    }
+                }
+            }
+
+            if dialer_done && listener_done {
+                return;
+            }
+        }
+    }
+
+    async fn listen_on_random_memory_address(&mut self) -> Multiaddr {
+        let memory_addr_listener_id = self.listen_on(get_rand_memory_address()).unwrap();
+
+        // block until we are actually listening
+        let multiaddr = loop {
+            match self.select_next_some().await {
+                SwarmEvent::NewListenAddr {
+                    address,
+                    listener_id,
+                } if listener_id == memory_addr_listener_id => {
+                    break address;
+                }
+                other => {
+                    log::debug!(
+                        "Ignoring {:?} while waiting for listening to succeed",
+                        other
+                    );
+                }
+            }
+        };
+
+        // Memory addresses are externally reachable because they all share the same memory-space.
+        self.add_external_address(multiaddr.clone(), AddressScore::Infinite);
+
+        multiaddr
+    }
+
+    async fn next_within(
+        &mut self,
+        seconds: u64,
+    ) -> SwarmEvent<<Self::NB as NetworkBehaviour>::OutEvent, THandlerErr<Self::NB>> {
+        match futures::future::select(
+            futures_timer::Delay::new(Duration::from_secs(seconds)),
+            self.select_next_some(),
+        )
+        .await
+        {
+            Either::Left(((), _)) => panic!("Swarm did not emit an event within {seconds}s"),
+            Either::Right((event, _)) => {
+                log::trace!("Swarm produced: {:?}", event);
+
+                event
+            }
+        }
+    }
+
+    async fn loop_on_next(mut self) {
+        while let Some(event) = self.next().await {
+            log::debug!("Swarm produced: {:?}", event);
+        }
+    }
+}
+
+fn get_rand_memory_address() -> Multiaddr {
+    let address_port = rand::random::<u64>();
+    let addr = format!("/memory/{}", address_port)
+        .parse::<Multiaddr>()
+        .unwrap();
+
+    addr
```

Comment:
```suggestion
    Protocol::Memory(0).into()
```
The memory transport picks a random port if the port is 0.

Reply:
That is so much clearer, thank you!